src/build_rock/configure/generate_build_matrix.py

```diff
@@ -0,0 +1,139 @@
+#!/usr/bin/env python3
+
+import yaml
+import os
+import argparse
+import json
+from enum import Enum
+from typing import Optional
+from ...shared import github_output
+
+
+class MATRIX_NAMES(Enum):
+    RUNNER = "runner-build-matrix"
+    LPCI = "lpci-build-matrix"
+
+
+class MissingArchSupport(Exception):
+    pass
+
+
+def get_target_archs(rockcraft: dict) -> list:
+    """get list of target architectures from rockcraft project definition"""
+
+    rock_platforms = rockcraft["platforms"]
+
+    target_archs = set()
+
+    for platf, values in rock_platforms.items():
+
+        if isinstance(values, dict) and "build-for" in values:
+            # TODO: Should this be "build-on"?
+            # https://documentation.ubuntu.com/rockcraft/en/latest/reference/rockcraft.yaml/#platforms-entry-build-on
+            if isinstance(arches := values["build-for"], list):
+                target_archs.update(arches)
+            elif isinstance(values, str):
+                target_archs.add(arches)
+        else:
+            target_archs.add(platf)
+
+    return target_archs
+
+
+def configure_matrices(target_archs: list, arch_map: dict, lp_fallback: bool) -> dict:
+    """Sort build into appropriate build matrices"""
+
+    # map configuration to individual job matrices
+    build_matrices = {name.value: {"include": []} for name in MATRIX_NAMES}
+
+    # Check if we have runners for all supported architectures
+    if missing_archs := set(target_archs) - set(arch_map.keys()):
+
+        # raise exception if we cannot fallback to LP builds
+        if not lp_fallback:
+            raise MissingArchSupport(
+                f"Missing support for runner arches: {missing_archs}"
+            )
+
+        # configure LP build
+        build_matrices[MATRIX_NAMES.LPCI.value]["include"].append(
+            {"architecture": "-".join(set(target_archs))}
```

Comment:
you would want to build only `missing_archs` with lpci, not all of them, right?

Reply:
That is a good question! I saw this earlier and I left this for later optimization. AFAIK this is the original behavior if you checkout the ppc64el-amd64 build logs here: https://github.com/canonical/oci-factory/actions/runs/10072927752/job/27849134914#logs

Do we need to build on the runner or lcpi exclusively?

Reply:
Actually, lpci currently builds for amd64 and ppc64le, although amd64 is already being built with github runners. I believe Cris did that so that in the case where the ppc64le build fails we can compare it the lpci's amd64 build and see what went wrong. Github runners only support amd64 and arm64, so lpci was used to build extra architectures, but it is going to be replaced soon by launchpad builds.

Reply:
If we are going to change it soon, maybe we should leave this behavior as it works in the current context.

Reply:
Indeed.